### PR TITLE
introduce ErrGroup, based on WaitGroup

### PR DESCRIPTION
**Pull request by amery**
and clean WaitGroup a bit

### CODE REVIEW

**Comment by amery** [comment]
I'm struggling to find a way to expose the cancellable context, and it's hard to justify passing the context to the shutdown function either.

**Comment by amery** [comment]
- use `.Parent` for the parent context instead of `.Context`
- export `eg.ctx` via `eg.Context()`
- introduce `eg.GoCatch()` to allow custom error catchers

File: sync.go [hunk 1]
```diff
@@ -112,3 +119,175 @@ func (wg *WaitGroup) Err() error {
...
+
+	eg.doGoCatch(r2, c2)
+}
+
+func (eg *ErrGroup) doGoCatch(run func() error, catch func(error) error) {
+	if catch == nil {
+		// default error catcher
+		catch = func(err error) error {
+			if err != nil && eg.Cancelled() {
+				err = os.ErrClosed
```
Comment (amery) [inline sync.go:285]: `err = context.Canceled` maybe?

**Review by karasz — APPROVED** [review]
Don't expect much comments on this except maybe:
```go
// Done returns a channel marker to know when the Group has been
// cancelled.
```
looks weird, `Done` is when `canceled`?

**Comment by amery** [comment]
> Don't expect much comments on this except maybe:
> 
> ```go
> // Done returns a channel marker to know when the Group has been
> // cancelled.
> ```
> 
> looks weird, `Done` is when `canceled`?

same as with `context.Context`, `ctx.Done()`'s channel gets closed with the cancellable context is cancelled

**Comment by amery** [comment]
`eg.Cancelled() bool` renamed to `eg.IsCancelled() bool` and `eg.Done() <-chan struct{}` to `eg.Cancelled() <=chan struct{}`

**Comment by amery** [comment]
> > Don't expect much comments on this except maybe:
> > ```go
> > // Done returns a channel marker to know when the Group has been
> > // cancelled.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > looks weird, `Done` is when `canceled`?
> 
> same as with `context.Context`, `ctx.Done()`'s channel gets closed with the cancellable context is cancelled.
> 
> @karasz should I add a note advising people to call `ctx.Wait()` after checking for `ctx.Done()`?

@karasz `Cancelled() bool` renamed to `IsCancelled() bool` and `Done() <-chan struct{}` to `Cancelled() >-chan struct{}` to make it clearer

**Comment by amery** [comment]
reduced `ErrGroup.GoCatch()` complexity

**Review by karasz — APPROVED** [review]
LGTM, although I think that for `core` we should have 100% unit test coverage, let's add that to a todo list

**Comment by amery** [comment]
> LGTM, although I think that for `core` we should have 100% unit test coverage, let's add that to a todo list

I fully agree, I just don't know how :crying_cat_face:

**Comment by amery** [comment]
`v0.11.0` tagged

**Pull request merged**